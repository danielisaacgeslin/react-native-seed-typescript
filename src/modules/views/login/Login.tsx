import React, { useEffect, useState, memo, useCallback } from 'react';
import { View, Text, TextInput, KeyboardAvoidingView } from 'react-native';
import { NavigationScreenProp } from 'react-navigation';
import { Button } from 'react-native-elements';

import { ENV, STYLE } from '../../../constants';
import styles from './styles';

export interface ILoginProps {
  isLoading: boolean;
  hasError: boolean;
  navigation: NavigationScreenProp<any, any>;
  isAndroid?: boolean;
  login: (email: string, password: string) => void;
  setNavigation: (navigation: NavigationScreenProp<any, any>) => void;
  checkForUpdates: () => void;
}

export interface ILoginState {
  email: string;
  password: string;
}

const Login = ({ navigation, hasError, isLoading, login, setNavigation, checkForUpdates, isAndroid = ENV.PLATFORM.IS_ANDROID }: ILoginProps) => {
  const [state, setState] = useState<ILoginState>({ email: 'fake@email.com', password: 'password' });

  const onChangeEmail = useCallback((email: string) => setState({ ...state, email }), [state, setState]);
  const onChangePassword = useCallback((password: string) => setState({ ...state, password }), [state, setState]);
  const onLogin = useCallback(() => login(state.email, state.password), [login, state]);

  useEffect(() => {
    setNavigation(navigation);
    checkForUpdates();
  });

  return (
    <View style={styles.container}>
      <KeyboardAvoidingView behavior="padding" style={styles.keyboardContainer} keyboardVerticalOffset={isAndroid ? ENV.KEYBOARD_VERTICAL_OFFSET : 0}>
        {hasError ? <Text style={styles.headingError}>Oops! Invalid email or password</Text> : null}
        <Text style={styles.heading1}>React Native Seed</Text>
        <Text style={styles.heading2}>Please login to your account</Text>
        <View style={styles.inputContainer}>
          <Text style={styles.inputLabel}>EMAIL:</Text>
          <TextInput
            keyboardType="email-address"
            placeholder="Enter your email..."
            placeholderTextColor={STYLE.COLOR.OCTANARY}
            style={[styles.input, hasError && styles.inputError]}
            underlineColorAndroid="rgba(0, 0, 0, 0)"
            value={state.email}
            onChangeText={onChangeEmail}
            testID="email-input"
          />
        </View>
        <View style={styles.inputContainer}>
          <Text style={styles.inputLabel}>PASSWORD</Text>
          <TextInput
            placeholder="Enter your password..."
            placeholderTextColor={STYLE.COLOR.OCTANARY}
            secureTextEntry={true}
            style={[styles.input, hasError && styles.inputError]}
            underlineColorAndroid="rgba(0, 0, 0, 0)"
            value={state.password}
            onChangeText={onChangePassword}
            testID="password-input"
          />
        </View>
        <Button loading={isLoading} disabled={isLoading} onPress={onLogin} title="Login" testID="login-button" disabledStyle={styles.buttonDisabled} />
      </KeyboardAvoidingView>
    </View>
  );
};

export default memo(Login);
