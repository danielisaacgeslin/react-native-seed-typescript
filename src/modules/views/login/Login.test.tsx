import React from 'react';
import Login, { ILoginProps } from './Login';
import { mount } from 'enzyme';

import { create } from 'react-test-renderer';

describe('Login', () => {
  let props: ILoginProps;
  let wrapper;

  global.console.error = () => null;
  beforeEach(() => {
    props = {
      isLoading: false,
      hasError: false,
      login: jest.fn(),
      setNavigation: jest.fn(),
      navigation: jest.fn() as any,
      checkForUpdates: jest.fn()
    };
    wrapper = mount(<Login {...props} />);
  });

  it('should render', () => {
    props.isAndroid = false;
    wrapper = mount(<Login {...props} />);
    expect(create(wrapper).toJSON()).toMatchSnapshot();
    expect(props.setNavigation).toBeCalledWith(props.navigation);
    expect(props.checkForUpdates).toBeCalled();
  });

  it('should render on Android', () => {
    props.isAndroid = true;
    wrapper = mount(<Login {...props} />);
    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });

  it('should render with errors', () => {
    props.hasError = true;
    wrapper = mount(<Login {...props} />);
    expect(create(wrapper).toJSON()).toMatchSnapshot();
  });

  it('should login', () => {
    const email = 'mail@mail.mail';
    const password = 'psw';
    wrapper
      .find('[testID="email-input"]')
      .first()
      .props()
      .onChangeText(email);
    wrapper.update();
    wrapper
      .find('[testID="password-input"]')
      .first()
      .props()
      .onChangeText(password);
    wrapper.update();
    wrapper
      .find('[testID="login-button"]')
      .first()
      .props()
      .onPress();
    expect(props.login).lastCalledWith(email, password);
  });
});
