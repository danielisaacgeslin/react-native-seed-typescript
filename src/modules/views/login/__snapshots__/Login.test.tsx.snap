// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "paddingBottom": 20,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#1745D5",
          "fontSize": 23,
          "fontWeight": "600",
          "marginBottom": 8,
          "marginTop": 16,
          "textAlign": "left",
        }
      }
    >
      Seed RN Typescript
    </Text>
    <Text
      style={
        Object {
          "color": "#9FA7BC",
          "fontSize": 17,
          "fontWeight": "400",
          "letterSpacing": 0.2,
          "marginBottom": 30,
        }
      }
    >
      Please login to your account
    </Text>
    <View
      style={
        Object {
          "marginBottom": 32,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9FA7BC",
            "fontSize": 13,
            "fontWeight": "600",
            "marginBottom": 10,
          }
        }
      >
        EMAIL:
      </Text>
      <TextInput
        allowFontScaling={true}
        keyboardType="email-address"
        onChangeText={[Function]}
        placeholder="Enter your email..."
        placeholderTextColor="#D5DEF1"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "#D5DEF1",
              "borderRadius": 4,
              "borderWidth": 1,
              "color": "#646565",
              "fontSize": 15,
              "fontWeight": "400",
              "padding": 16,
            },
            false,
          ]
        }
        testID="email-input"
        underlineColorAndroid="rgba(0, 0, 0, 0)"
        value="fake@email.com"
      />
    </View>
    <View
      style={
        Object {
          "marginBottom": 32,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9FA7BC",
            "fontSize": 13,
            "fontWeight": "600",
            "marginBottom": 10,
          }
        }
      >
        PASSWORD
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Enter your password..."
        placeholderTextColor="#D5DEF1"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#D5DEF1",
              "borderRadius": 4,
              "borderWidth": 1,
              "color": "#646565",
              "fontSize": 15,
              "fontWeight": "400",
              "padding": 16,
            },
            false,
          ]
        }
        testID="password-input"
        underlineColorAndroid="rgba(0, 0, 0, 0)"
        value="password"
      />
    </View>
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="login-button"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Login
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Login should render on Android 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "paddingBottom": 20,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#1745D5",
          "fontSize": 23,
          "fontWeight": "600",
          "marginBottom": 8,
          "marginTop": 16,
          "textAlign": "left",
        }
      }
    >
      Seed RN Typescript
    </Text>
    <Text
      style={
        Object {
          "color": "#9FA7BC",
          "fontSize": 17,
          "fontWeight": "400",
          "letterSpacing": 0.2,
          "marginBottom": 30,
        }
      }
    >
      Please login to your account
    </Text>
    <View
      style={
        Object {
          "marginBottom": 32,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9FA7BC",
            "fontSize": 13,
            "fontWeight": "600",
            "marginBottom": 10,
          }
        }
      >
        EMAIL:
      </Text>
      <TextInput
        allowFontScaling={true}
        keyboardType="email-address"
        onChangeText={[Function]}
        placeholder="Enter your email..."
        placeholderTextColor="#D5DEF1"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "#D5DEF1",
              "borderRadius": 4,
              "borderWidth": 1,
              "color": "#646565",
              "fontSize": 15,
              "fontWeight": "400",
              "padding": 16,
            },
            false,
          ]
        }
        testID="email-input"
        underlineColorAndroid="rgba(0, 0, 0, 0)"
        value="fake@email.com"
      />
    </View>
    <View
      style={
        Object {
          "marginBottom": 32,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9FA7BC",
            "fontSize": 13,
            "fontWeight": "600",
            "marginBottom": 10,
          }
        }
      >
        PASSWORD
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Enter your password..."
        placeholderTextColor="#D5DEF1"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#D5DEF1",
              "borderRadius": 4,
              "borderWidth": 1,
              "color": "#646565",
              "fontSize": 15,
              "fontWeight": "400",
              "padding": 16,
            },
            false,
          ]
        }
        testID="password-input"
        underlineColorAndroid="rgba(0, 0, 0, 0)"
        value="password"
      />
    </View>
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="login-button"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Login
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Login should render with errors 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "paddingBottom": 20,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "paddingHorizontal": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#C21E5D",
          "fontSize": 13,
          "fontWeight": "400",
          "marginTop": 8,
        }
      }
    >
      Oops! Invalid email or password
    </Text>
    <Text
      style={
        Object {
          "color": "#1745D5",
          "fontSize": 23,
          "fontWeight": "600",
          "marginBottom": 8,
          "marginTop": 16,
          "textAlign": "left",
        }
      }
    >
      Seed RN Typescript
    </Text>
    <Text
      style={
        Object {
          "color": "#9FA7BC",
          "fontSize": 17,
          "fontWeight": "400",
          "letterSpacing": 0.2,
          "marginBottom": 30,
        }
      }
    >
      Please login to your account
    </Text>
    <View
      style={
        Object {
          "marginBottom": 32,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9FA7BC",
            "fontSize": 13,
            "fontWeight": "600",
            "marginBottom": 10,
          }
        }
      >
        EMAIL:
      </Text>
      <TextInput
        allowFontScaling={true}
        keyboardType="email-address"
        onChangeText={[Function]}
        placeholder="Enter your email..."
        placeholderTextColor="#D5DEF1"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "#D5DEF1",
              "borderRadius": 4,
              "borderWidth": 1,
              "color": "#646565",
              "fontSize": 15,
              "fontWeight": "400",
              "padding": 16,
            },
            Object {
              "borderColor": "#C21E5D",
              "borderWidth": 1,
            },
          ]
        }
        testID="email-input"
        underlineColorAndroid="rgba(0, 0, 0, 0)"
        value="fake@email.com"
      />
    </View>
    <View
      style={
        Object {
          "marginBottom": 32,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#9FA7BC",
            "fontSize": 13,
            "fontWeight": "600",
            "marginBottom": 10,
          }
        }
      >
        PASSWORD
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Enter your password..."
        placeholderTextColor="#D5DEF1"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "borderColor": "#D5DEF1",
              "borderRadius": 4,
              "borderWidth": 1,
              "color": "#646565",
              "fontSize": 15,
              "fontWeight": "400",
              "padding": 16,
            },
            Object {
              "borderColor": "#C21E5D",
              "borderWidth": 1,
            },
          ]
        }
        testID="password-input"
        underlineColorAndroid="rgba(0, 0, 0, 0)"
        value="password"
      />
    </View>
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="login-button"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Login
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
